src/client/pydaos/pydaos_shim.c

```diff
@@ -446,6 +446,273 @@ __shim_handle__cont_close(PyObject *self, PyObject *args)
+{
+	char			*buf = NULL;
+	daos_anchor_t		anchor = {0};
+	daos_key_desc_t		kds[ITER_NR];
+	d_sg_list_t		sgl;
+	d_iov_t			sg_iov;
+	daos_size_t		buf_size = ITER_NR * 256;
+	struct pydaos_df	entry;
+	size_t			size = sizeof(entry);
+	d_iov_t			marker;
+	bool			mark_data = true;
+	int			rc = 0;
+
+	D_ALLOC(buf, buf_size);
+	if (buf == NULL)
+		goto out;
+
+	d_iov_set(&marker, &mark_data, sizeof(mark_data));
+	d_iov_set(&sg_iov, buf, buf_size);
+	sgl.sg_nr		= 1;
+	sgl.sg_nr_out		= 0;
+	sgl.sg_iovs		= &sg_iov;
+
+	while (!daos_anchor_is_eof(&anchor)) {
+		uint32_t	nr = ITER_NR, i;
+		void		*ptr;
+
+		memset(buf, 0, buf_size);
+		rc = daos_kv_list(oh, DAOS_TX_NONE, &nr, kds, &sgl, &anchor, NULL);
+		if (rc)
+			goto out;
+		if (nr == 0)
+			continue;
+
+		for (ptr = buf, i = 0; i < nr; i++) {
+			char *key;
+
+			D_ALLOC(key, kds[i].kd_key_len + 1);
+			if (key == NULL) {
+				rc = -DER_NOMEM;
+				goto out;
+			}
+
+			memcpy(key, ptr, kds[i].kd_key_len);
+			key[kds[i].kd_key_len + 1] = '\0';
+			ptr += kds[i].kd_key_len;
+
+			rc = daos_kv_get(oh, DAOS_TX_NONE, DAOS_COND_KEY_GET, key, &size, &entry,
+					 NULL);
+			D_FREE(key);
+			if (rc)
+				goto out;
+
+			rc = daos_oit_mark(oit, entry.oid, &marker, NULL);
+			if (rc) {
+				D_ERROR("Failed to mark OID in OIT: "DF_RC"\n", DP_RC(rc));
+				goto out;
+			}
+		}
+	}
+
+out:
+	D_FREE(buf);
+	return rc;
+}
+
+static PyObject *
+cont_check(int ret, char *pool, char *cont, int flags)
+{
+	daos_handle_t			coh = {0};
+	daos_handle_t			poh = {0};
+	daos_handle_t			oh = {0};
+	daos_handle_t			oit = {0};
+	daos_prop_t			*prop = NULL;
+	struct daos_prop_entry		*entry;
+	struct daos_prop_co_roots	*roots;
+	daos_epoch_t			snap_epoch = 0;
+	daos_anchor_t			anchor = {0};
+	uint32_t			nr_entries = ITER_NR, i;
+	daos_obj_id_t			oids[ITER_NR] = {0};
+	daos_epoch_range_t		epr;
+	d_iov_t				marker;
+	bool				mark_data = true;
+	int				rc, rc2;
+
+	if (ret != DER_SUCCESS) {
+		rc = ret;
+		goto out;
+	}
+
+	/** Connect to pool */
+	rc = daos_pool_connect(pool, NULL, DAOS_PC_RW, &poh, NULL, NULL);
+	if (rc)
+		goto out;
+
+	/** Open container. TODO - need exclusive open when available */
+	rc = daos_cont_open(poh, cont, DAOS_COO_RW, &coh, NULL, NULL);
+	if (rc)
+		goto out;
+
+	/** create snapshot for OIT */
+	rc = daos_cont_create_snap_opt(coh, &snap_epoch, NULL, DAOS_SNAP_OPT_CR | DAOS_SNAP_OPT_OIT,
+				       NULL);
+	if (rc)
+		goto out;
+
+	/** Open OIT table */
+	rc = daos_oit_open(coh, snap_epoch, &oit, NULL);
+	if (rc)
+		goto out;
+
+	/** Retrieve container properties via cont_query() */
+	prop = daos_prop_alloc(0);
+	if (prop == NULL) {
+		rc = -DER_NOMEM;
+		goto out;
+	}
+
+	rc = daos_cont_query(coh, NULL, prop, NULL);
+	if (rc)
+		goto out;
+
+	/** Verify that this is a python container */
+	entry = daos_prop_entry_get(prop, DAOS_PROP_CO_LAYOUT_TYPE);
+	if (entry == NULL || entry->dpe_val != DAOS_PROP_CO_LAYOUT_PYTHON) {
+		D_ERROR("Container is not a python container\n");
+		rc = -DER_INVAL;
```

Comment:
D_ERROR calls should include DF_RC of the rc value.

Reply:
this really should have been caught and change requested before the change landed to the feature branch. i don't see a need to repush this for that. will create a ticket to address those later.